pkg/p2p/libp2p/connections_test.go

```diff
@@ -317,3 +321,130 @@ func TestConnectRepeatHandshake(t *testing.T) {
 	expectPeersEventually(t, s2)
 	expectPeersEventually(t, s1)
 }
+
+func TestTopologyNotifiee(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	var (
+		mtx                sync.Mutex
+		n1connectedAddr    swarm.Address
+		n1disconnectedAddr swarm.Address
+		n2connectedAddr    swarm.Address
+		n2disconnectedAddr swarm.Address
+
+		n1c = func(_ context.Context, a swarm.Address) error {
+			mtx.Lock()
+			defer mtx.Unlock()
+			expectZeroAddress(t, n1connectedAddr) // fail if set more than once
+			n1connectedAddr = a
+			return nil
+		}
+		n1d = func(a swarm.Address) {
+			mtx.Lock()
+			defer mtx.Unlock()
+			n1disconnectedAddr = a
+		}
+
+		n2c = func(_ context.Context, a swarm.Address) error {
+			mtx.Lock()
+			defer mtx.Unlock()
+			expectZeroAddress(t, n2connectedAddr) // fail if set more than once
+			n2connectedAddr = a
+			return nil
+		}
+		n2d = func(a swarm.Address) {
+			mtx.Lock()
+			defer mtx.Unlock()
+			n2disconnectedAddr = a
+		}
+	)
+	s1, overlay1, cleanup1 := newService(t, libp2p.Options{NetworkID: 1, Notifiee: mockNotifiee(n1c, n1d)})
+	defer cleanup1()
+	s2, overlay2, cleanup2 := newService(t, libp2p.Options{NetworkID: 1, Notifiee: mockNotifiee(n2c, n2d)})
+	defer cleanup2()
+
+	addr := serviceUnderlayAddress(t, s1)
+
+	// s2 connects to s1, thus the notifiee on s1 should be called on Connect
+	overlay, err := s2.Connect(ctx, addr)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	expectPeers(t, s2, overlay1)
+	expectPeersEventually(t, s1, overlay2)
+
+	// expect that n1 notifee called with s2 overlay
+	waitAddrSet(t, &n1connectedAddr, &mtx, overlay2)
+
+	mtx.Lock()
+	expectZeroAddress(t, n1disconnectedAddr, n2connectedAddr, n2disconnectedAddr)
+	mtx.Unlock()
+
+	// s2 disconnects from s1 so s1 disconnect notifiee should be called
+	if err := s2.Disconnect(overlay); err != nil {
+		t.Fatal(err)
+	}
+
+	expectPeers(t, s2)
+	expectPeersEventually(t, s1)
+	waitAddrSet(t, &n1disconnectedAddr, &mtx, overlay2)
+
+	mtx.Lock()
+	expectZeroAddress(t, n2connectedAddr)
+	mtx.Unlock()
```

Comment:
i added a comment but will also include a more elaborate version here:
- when you connect to a peer, it is the topology that asks libp2p to connect to that peer, so if the connection is successful, the topology already knows it and adds the peer to the list of connected nodes
- when you disconnect from a peer, you can either disconnect from the topology, but actually any other protocol or component interacting with the peer could choose to break the connection with the peer (misbehaving fraudulent nodes that cheat on account/sending bogus chunks). this eliminates the need to introduce dependencies between those components and the actual topology component. if some different component wants to disconnect a peer it can do so in this case, resulting in the topology being notified about the disconnection
- so, when you disconnect from the node, topology would be notified (since it might not have initiated the disconnection)
- the other peer's topology would also get notified that you have disconnected from them
- therefore, 2 notifications in total

Reply:
neat, thanks